core/src/app.rs

```diff
@@ -299,6 +364,35 @@ impl App {
                     .collect::<Vec<_>>();
             }
 
+            // Special post-processing of while blocks, if not already in the while loop, mark that
+            // we're in a while loop by setting `current_while`. This means we'll aggregate block
+            // results as arrays instead of values in the `env.state`.
+            if block.block_type() == BlockType::While {
+                match current_while {
+                    Some(w) => {
+                        assert!(w == block_idx);
+                        current_while_iteration = Some(current_while_iteration.unwrap() + 1);
+                    }
+                    None => {
+                        current_while = Some(block_idx);
+                        current_skips = Some(envs.iter().map(|_| false).collect());
+                        current_while_iteration = Some(0);
+                    }
+                }
+                envs = envs
+                    .iter()
+                    .map(|map_envs| {
+                        assert!(map_envs.len() == 1);
+                        let mut env = map_envs[0].clone();
+                        env.map = Some(MapState {
+                            name: name.clone(),
+                            iteration: current_while_iteration.unwrap(),
```

Comment:
Yeah there is a lot of invariants I rely on here and this is one of them. Expect could be better or maching on it not being None, but that creates more code. But ideally we would have proper errors instead of asserts or unwrap I agree.

TBH I expect some of these to be possibly slightly incorrect in which case we'll hit asserts that we can fix later on.